### ISSUE TITLE
Implement StagingArea instead of Queues

### ISSUE THREAD

**Issue by proteneer** [deepchem/deepchem#789]
With deterministic now often set to False when training, it makes sense for us to consider using StagingAreas instead of FIFOQueues for performance critical models (eg. ANI). 

https://www.tensorflow.org/api_docs/python/tf/contrib/staging/StagingArea

**Comment by proteneer** [comment]
Note that StagingArea.clear() and capacity abilities are only available in tensorflow 1.3 - what's our roadmap for upgrading to 1.3?

**Comment by rbharath** [comment]
How big a perf difference would switching to StagingAreas make?

I'll open a new issue to coordinate the 1.3 upgrade.

**Comment by rbharath** [comment]
Started #790 for tracking 1.3 upgrade plan.

**Comment by proteneer** [comment]
Quite a lot - I estimate about 10% or more based on my current profiling
![Profiler](http://i.imgur.com/ijH59Yf.png)

See the first 5ms - it looks like it's all pulling data from the FIFOQueue. In this case it's about 16% of the total run time.

**Comment by proteneer** [comment]
Note that we may be able to get away with tf 1.2, it has a more limited version of StagingArea (lack of capacity and clear()). We will also need to redo our threading/locks since now it will block, the FIFOQueue equivalent of calling close() and sending an exception does not apply here. Instead, pulling from StagingArea will block.

**Comment by proteneer** [comment]
An alternative is to try and use enqueue/dequeue_many, but we're already pretty batch_size limited already so I don't know if adding more batches into the queue will help.

**Comment by proteneer** [comment]
This is how it looks with StagingArea (note that I'm calling it FIFOQueue still, but it's actually using get() and put() ops from staging area)

![Staging](http://i.imgur.com/JJ4kq2I.png)

**Comment by rbharath** [comment]
I'd generally be open to upgrading, but we should make sure that performance is actually increasing with some careful benchmarking work up front.

**Comment by peastman** [comment]
What's the status of this?  I want to look for ways to speed up TensorGraph, and this looks like one of the top candidates.

**Comment by proteneer** [comment]
@peastman I have a barely working/super buggy version of this. The main reason I haven't pushed it yet is because StagingAreas areas are fully blocking and are absolutely non-deterministic. I got a 10-15% speed up with it, you might give it a shot and see how it performs on your system.

On the tf 1.3 side of things, I'm open to upgrading.

**Comment by peastman** [comment]
Can you post your implementation or send it to me?  I'd love to take a look at it.

**Comment by proteneer** [comment]
@peastman check PR #845

**Comment by peastman** [comment]
I've copied over your changes, plus added some synchronization that should be very lightweight so it doesn't hang at the end.  I'm not seeing any speed improvement, and if anything it seems to be slightly slower.  Here's the test I'm using.

```python
import numpy as np
import deepchem as dc

muv_tasks, muv_datasets, transformers = dc.molnet.load_muv()
train_dataset, valid_dataset, test_dataset = muv_datasets 

model = dc.models.TensorGraphMultiTaskClassifier(
    len(muv_tasks), n_features=1024, dropouts=[.25],
    learning_rate=0.001, weight_init_stddevs=[.1],
    batch_size=64)

model.fit(train_dataset, nb_epoch=10)
```

With FIFOQueue fitting takes 35 seconds.  With StagingArea, it's 36 seconds.  With the queue disabled, it's 40 seconds.

Do you have a test that shows a difference?  Why do you expect StagingArea to be faster than FIFOQueue?

**Comment by proteneer** [comment]
Interesting.

I'm using the ANI model (viewable via my PR). I pinpointed the FIFOQueue op when I did a timelines profile. The NN itself is quite simple, so it's possible the data IO was more exacerbated. It is weird that SatgingArea made things slower though. Are you sure the threadig implementation was lightweight enough? (Try having it block so you don't have to worry about the threading overhead).

As for why I chose StagingArea to start with:

https://www.tensorflow.org/versions/r1.3/performance/performance_models

"The dominant part of each stage is executed in parallel with the other stages using data_flow_ops.StagingArea. StagingArea is a queue-like operator similar to tf.FIFOQueue. The difference is that StagingArea offers simpler functionality and can be executed on both CPU and GPU in parallel with other stages. Breaking the input pipeline into 3 stages that operate independently in parallel is scalable and takes full advantage of large multi-core environments. The rest of this section details the stages followed by details about using data_flow_ops.StagingArea."

**Comment by peastman** [comment]
Here's how I did the synchronization.  I start by creating two lists (really just references to two numbers) for communicating between the threads:

```python
n_enqueued = [0]
final_sample = [None]
```

The enqueue thread increments `n_enqueued[0]` for each batch, and sets `final_sample[0] = n_enqueued[0]` when it's done.  The main thread executes the following at the start of each iteration:

```python
if self.use_queue:
  # Don't let this thread get ahead of the enqueue thread, since if
  # we try to read more batches than the total number that get queued,
  # this thread will hang indefinitely.
  while n_enqueued[0] <= n_samples:
    if n_samples == final_sample[0]:
      break
    time.sleep(0)
  if n_samples == final_sample[0]:
    break
```

I verified that `time.sleep(0)` (which is Python's idiom for forcing a thread to yield) only gets executed very infrequently: about 11,000 times during over 6 million steps of fitting.

The paragraph you quoted may not be relevant.  Here are the three stages they're using StagingArea to coordinate:

> - I/O reads: Choose and read image files from disk.
> - Image Processing: Decode image records into images, preprocess, and organize into mini-batches.
> - CPU-to-GPU Data Transfer: Transfer images from CPU to GPU.

They're letting TensorFlow handle all their disk I/O, which doesn't involve the GPU at all.  That's why they need something that can execute on the CPU and GPU in parallel.  That doesn't apply to us.

**Comment by peastman** [comment]
I just tried another test: do synchronization as described above, but with a FIFOQueue instead of a StagingArea.  That didn't make any difference.  I'm not sure the StagingArea is actually slower: the times fluctuate a bit, and the difference is likely within the noise level.  But it certainly doesn't seem to be any faster.

**Comment by peastman** [comment]
Regardless of which class we use, it might be a good idea to do the synchronization this way.  The current implementation throws an exception if you call `fit()` more than once on a model.  That will be a problem for GANs where you want to call `fit()` repeatedly (see #535).

**Comment by rbharath** [comment]
I'm going to close this since I think `tf.data` is the new blessed way to do high-performance TensorFlow models. Let's discuss over at #1039.
